website/docs/developer/research/research-lambda-container.md

```diff
@@ -0,0 +1,144 @@
+
+1. Create an ECR repository from AWS CLI if needed as shown. Check [here](https://docs.aws.amazon.com/AmazonECR/latest/userguide/getting-started-cli.html#cli-create-repository) for additional details on this.
+
+```bash
+aws ecr create-repository \
+    --repository-name <YOUR_REPOSITORY_NAME>
+```
+
+2. Create a project directory. Under that directory, add  your script (`test.py` in this case) and `requirements.txt`to install any dependencies. Then  create a `Dockerfile` that creates the image. An example of a Dockerfile used for prototying a lambda having `test.py` file is shown below.
+
+```yaml
+FROM public.ecr.aws/lambda/python:3.8
+
+# Copy function code
+COPY test.py ${LAMBDA_TASK_ROOT}
+
+# Install the function's dependencies using file requirements.txt
+# from your project folder.
+
+COPY requirements.txt  .
+RUN  pip3 install -r requirements.txt --target "${LAMBDA_TASK_ROOT}"
+
+# Set the CMD to your handler (could also be done as a parameter override outside of the Dockerfile)
+CMD [ "test.hambda_handler" ]
+
+```
+
+3. The next steps are to build, tag and push the image to the ECR repo. You can build a new image named `prototype-lambda-image` in this case using:
+
+```bash
+docker build -t prototype-lambda-image .
+```
+Once build is successful, tag the image
+
+```bash
+docker tag prototype-lambda-image:<IMAGE_TAG> <YOUR_ECR_REPO_URI>:<IMAGE_TAG>
+```
+Next, login to the ECR repo using:
+
+```bash
+aws ecr get-login-password --region <YOUR_REGION> | docker login --username AWS --password-stdin <YOUR_AWS_ACCOUNT_ID>.dkr.ecr.<YOUR_REGION>.amazonaws.com
+```
+Check this [link](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html) for additional information on pushing image to ECR using Docker CLI.
+
+Finally, push the image to ECR using:
+
+```bash
+docker push <YOUR_ECR_REPO_URI>:<IMAGE_TAG>
+```
+
+4. The next step is to deploy the lambda using terraform. Use the following example code to deploy the lambda container:
+
+```terraform
+
+resource "aws_lambda_function" "prototype_lambda" {
+  image_uri     = "<YOUR_ECR_REPO_URI>:<IMAGE_TAG>"  # repo and tag
+  package_type  = "Image"
+  function_name = "prototype-lambda"
+  role          = "<YOUR_IAM_ROLE_ARN>"
+  image_config {
+    command = ["test.lambda_handler"]
+  }
+}
+```
+Using the steps above, a prototype has been created in NGAP AWS sandbox account which can be seen under the lambda function console. The lambda function is named as `prototype-lambda-container` and uses the `test:latest` private ECR repo image.
+
+
+## Future directions
+Currently, the only option to store the Docker image for lambda containers is AWS ECR repository. If github is supported in the future, then using that approach will be a good direction to go.
+
+A few possible discussion items:
+- Discuss with NGAP team if the docker images can be stored in a public ECR repo.
+- Check if github can be used to store the image (currently I have not seen any example online)
```

Comment:
if images are stored in github instead of ECR, how can it be used to deploy the lambda?

as you can see in the config files of the prototype I created, you will need ECR repo link to add in the lambda terraform.

```resource "aws_lambda_function" "prototype_lambda" {
  image_uri     = "<YOUR_ECR_REPO_URI>:<IMAGE_TAG>"  # repo and tag
```

Reply:
So currently, they can store images but they cant be used for deployment because of the AWS lock down on where images come from for Lambdas. In fact you state that earlier in the document that an ECR mst be used. I misunderstood the bullet. I would maybe rewrite the bullet to discuss how to handle storage of the images in a private repo for each user if a public repo is out of the question. Maybe how we automate builds for them or other methods.

Reply:
Added some info on this.

Also created [this card](https://bugs.earthdata.nasa.gov/browse/ORCA-375) for additional research on builds.